rllib/utils/nested_dict.py

```diff
@@ -0,0 +1,280 @@
+# Copyright 2021 DeepMind Technologies Limited.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Custom NestedDict datatype."""
+
+import itertools
+from typing import (
+    AbstractSet,
+    Any,
+    Dict,
+    Iterable,
+    Iterator,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+)
+
+
+StreamType = Union[str, Sequence[str]]
+_IndexType = Union[StreamType, Sequence[StreamType]]
+
+_NestedDictType = Dict[str, Any]
+_NestedMappingType = Mapping[_IndexType, Any]
+
+NestedDictInputType = Union[
+    Iterable[Tuple[StreamType, Any]], _NestedMappingType, "NestedDict"
+]
+
+
+def _flatten_index(index: _IndexType) -> Sequence[str]:
+    if isinstance(index, str):
+        return (index,)
+    else:
+        return tuple(itertools.chain.from_iterable([_flatten_index(y) for y in index]))
```

Comment:
We seem to call this a lot. Perhaps it should have a docstring and not be prefixed with an underscore? If I understand this correctly, it is basically `nested_dict.keys()` which I think developers would want to use.

Reply:
it's a private helper function for the NestedDict class. Noone should use this method for now.

Reply:
dude, if that's the case, you should name it __flatten_keys().
can happen with some other PRs.